server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java

```diff
@@ -152,21 +149,29 @@ protected boolean hasOrganizationPermissionImpl(String organizationUuid, String
 
   @Override
   protected boolean hasProjectUuidPermission(String permission, String projectUuid) {
-    if (!projectPermissionsCheckedByUuid.contains(permission)) {
-      try (DbSession dbSession = dbClient.openSession(false)) {
-        // FIXME do not load all the authorized projects. It can be huge.
-        Collection<String> projectUuids = dbClient.authorizationDao().selectAuthorizedRootProjectsUuids(dbSession, getUserId(), permission);
-        addProjectPermission(permission, projectUuids);
-      }
+    if (permissionsByProjectUuid == null) {
+      permissionsByProjectUuid = HashMultimap.create();
     }
-    return projectUuidByPermission.get(permission).contains(projectUuid);
+    Set<String> permissions;
+    if (permissionsByProjectUuid.containsKey(projectUuid)) {
+      permissions = permissionsByProjectUuid.get(projectUuid);
+    } else {
+      permissions = loadProjectPermissions(projectUuid);
+      permissionsByProjectUuid.putAll(projectUuid, permissions);
```

Comment:
according to javadoc of `putAll`: 
> In particular, this is a no-op if {@code values} is empty.

I'm wondering there is actually any caching when there is no permission for the projectUuid.
If it's a no-op, then the key won't be added to the Multimap and `containsKey` will always return `false`, hence no caching the next time method is called with the same `projectUuid`.

In which case, a `Map<String, Set<String>>` should be used instead (and the call to `containsKey` dropped in favor of a single call to `get`)